javascript/ql/src/semmle/javascript/NullSensitiveContext.qll

```diff
@@ -0,0 +1,488 @@
+ *
+ * The expression found in that context is called the principal argument of
+ * the context.
+ *
+ * In examples, we denote the position of the principal argument by use of
+ * an underscore. For example, in `_ + 0`, the context's principle argument
+ * is the left operand of the addition.
+ */
+abstract class NullSensitiveContext extends Expr {
+  abstract Expr getPrincipalArgument();
+}
+
+/*
+ * _[p]
+ */
+
+class IndexExprBase extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * o[_]
+ */
+
+class IndexExprIndex extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getPropertyNameExpr() }
+}
+
+/*
+ * _.p
+ */
+
+class DotExprBase extends NullSensitiveContext, DotExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * new _
+ */
+
+class NewExprCallee extends NullSensitiveContext, NewExpr {
+  override Expr getPrincipalArgument() { result = this.getCallee() }
+}
+
+/*
+ * _(args)
+ */
+
+class CallExprCallee extends NullSensitiveContext, CallExpr {
+  override Expr getPrincipalArgument() { result = this.getCallee() }
+}
+
+/*
+ * _ + y
+ */
+
+class AddExprLeftOperand extends NullSensitiveContext, AddExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x + _
+ */
+
+class AddExprRightOperand extends NullSensitiveContext, AddExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * _ - y
+ */
+
+class SubExprLeftOperand extends NullSensitiveContext, SubExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x - _
+ */
+
+class SubExprRightOperand extends NullSensitiveContext, SubExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * _ * u
+ */
+
+class MulExprLeftOperand extends NullSensitiveContext, MulExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x * _
+ */
+
+class MulExprRightOperand extends NullSensitiveContext, MulExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * _ / y
+ */
+
+class DivExprLeftOperand extends NullSensitiveContext, DivExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x / _
+ */
+
+class DivExprRightOperand extends NullSensitiveContext, DivExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * _ % y
+ */
+
+class ModExprLeftOperand extends NullSensitiveContext, ModExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x % _
+ */
+
+class ModExprRightOperand extends NullSensitiveContext, ModExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * +_
+ */
+
+class PlusExprOperand extends NullSensitiveContext, PlusExpr {
```

Comment:
Generalise to `extends UnaryExpr`, excluding `LogNotExpr` in the characteristic predicate.